master/buildbot/www/avatar.py

```diff
@@ -34,12 +37,143 @@ def getUserAvatar(self, email, username, size, defaultAvatarUrl):
+            # Not found
+            return None
+        if 200 <= res.code < 300:
+            data = yield res.json()
+            return data['avatar_url']
+
+        log.msg('Failed looking up user: response code {}'.format(res.code))
+        return None
 
+    @defer.inlineCallbacks
+    def _search_avatar_by_user_email(self, email):
+        headers = {
+            'Accept': 'application/vnd.github.v3+json',
+        }
+
+        query = '{} in:email'.format(email)
+        url = '/search/users?{}'.format(urlencode({
+            'q': query,
+        }))
+        http = yield self._getHTTPClient()
+        res = yield http.get(url, headers=headers)
+        if 200 <= res.code < 300:
+            data = yield res.json()
+            if data['total_count'] == 0:
+                # Not found
+                return None
+            return data['items'][0]['avatar_url']
+
+        log.msg('Failed searching user by email: response code {}'.format(res.code))
+        return None
+
+    @defer.inlineCallbacks
+    def _search_avatar_by_commit(self, email):
+        headers = {
+            'Accept': 'application/vnd.github.v3+json,application/vnd.github.cloak-preview',
+        }
+
+        query = {
+            'q': 'author-email:{}'.format(email),
+            'sort': 'committer-date',
+            'per_page': '1',
+        }
+        sorted_query = sorted(query.items(), key=lambda x: x[0])
+        url = '/search/commits?{}'.format(urlencode(sorted_query))
+        http = yield self._getHTTPClient()
+        res = yield http.get(url, headers=headers)
+        if 200 <= res.code < 300:
+            data = yield res.json()
+            if data['total_count'] == 0:
+                # Not found
+                return None
+            return data['items'][0]['author']['avatar_url']
+
+        log.msg('Failed searching user by commit: response code {}'.format(res.code))
+        return None
+
+    def _add_size_to_url(self, avatar, size):
+        parts = urlparse(avatar)
+        query = parts.query
+        if query:
+            query += '&'
+        query += 's={0}'.format(size)
+        return urlunparse((parts.scheme,
+            parts.netloc, parts.path, parts.params,
+            query, parts.fragment))
+
+    @defer.inlineCallbacks
+    def getUserAvatar(self, email, username, size, defaultAvatarUrl):
+        avatar = None
+        if username:
+            username = username.decode('utf-8')
+        if email:
+            email = email.decode('utf-8')
+
+        if username:
+            avatar = yield self._get_avatar_by_username(username)
+        if not avatar and email:
+            # Try searching a user with said mail
+            avatar = yield self._search_avatar_by_user_email(email)
+        if not avatar and email:
+            # No luck, try to find a commit with this email
+            avatar = yield self._search_avatar_by_commit(email)
+
+        if not avatar:
+            # No luck
+            return
```

Comment:
I think we should do `return None` to make code more explicit.

Reply:
Done